### ISSUE TITLE
Add graceful error handling to Git branch fetch

### ISSUE THREAD

**Issue by dagood** [dotnet/buildtools#1980]
If the current directory isn't a Git repo (such as during source-build's tarball build), this code ends up putting the Git output's newlines in a C++ string, which gives CoreCLR compile errors:

https://github.com/dotnet/buildtools/blob/2f0dec8eba53942a758910ea7b4cc79e4a7d2c36/src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets#L266-L269

It should handle error exit codes like it does when getting the commit:

https://github.com/dotnet/buildtools/blob/2f0dec8eba53942a758910ea7b4cc79e4a7d2c36/src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets#L271-L285

This will also let us pass in the branch name from source-build if we know it.

**Comment by weshaggard** [comment]
As part of this fix we should change the message importance from high to normal.

**Comment by vancem** [comment]
I don't understand why the original code was causing problems.  The reason for the 'RawGitBranchName was that I only use it as a branch name if it looks like output I expect (namely if it begins with a 'heads' or 'remotes/origin/'    
```
    <!-- If things match expected patterns, create $(GitBranchName) from $(RawGitBranchName) -->
    <Message Importance="high" Text="GIT Branch Name as seen by git describe: '$(RawGitBranchName)'" /> 
    <PropertyGroup Condition="$(RawGitBranchName.StartsWith('heads/'))">
      <GitBranchName>$(RawGitBranchName.SubString(6))</GitBranchName>
    </PropertyGroup>
    <PropertyGroup Condition="$(RawGitBranchName.StartsWith('remotes/origin/'))">
      <GitBranchName>$(RawGitBranchName.SubString(15))</GitBranchName>
    </PropertyGroup>
    <Message Importance="high" Text="GIT Branch Name: '$(GitBranchName)'" /> 
```
On errors, I would have expected it to not machine, and thus leave the GitBranchName empty (which means it should not be shown.

Presumably this did not happen, but I don't understand why my logic above is wrong...

**Comment by dagood** [comment]
I think you're actually right, and I misread the code... I hit this problem before your changes when the git command was getting `GitBranchName` directly, and I was thinking of the matching patterns you added as "extra" ways to try to get a meaningful branch name, rather than filters.

I tried a source-build tarball build of CoreCLR with a BuildTools that contains your change (and doesn't contain my workaround) and it works fine. Sorry about that. :smile:

If there are any error messages or multiline responses that can start with `heads/` or `remotes/origin/`, there's still a problem, but I don't think that's the case.

Closing as (already) fixed.
